src/main/java/com/imsweb/mph/DefaultHematoDbDataProvider.java

```diff
@@ -92,70 +91,24 @@ public DefaultHematoDbUtilsProvider() {
     }
 
     @Override
-    public MpResult isSamePrimary(String leftCode, String rightCode, int leftYear, int rightYear) {
-        if (leftCode == null || rightCode == null || !_MORPHOLOGY.matcher(leftCode).matches() || !_MORPHOLOGY.matcher(rightCode).matches())
-            return MpResult.INVALID_INPUT;
-        if (leftCode.equals(rightCode))
-            return MpResult.SINGLE_PRIMARY;
-
-        boolean rightCodeIsSamePrimaryForLeft = false;
-        if (_samePrimaryDto.containsKey(leftCode)) {
-            List<HematoDbDTO> allRowsForLeftCode = _samePrimaryDto.get(leftCode);
-            List<HematoDbDTO> validRowsForLeftCode = allRowsForLeftCode.stream().filter(r -> r.validCodeForDiagnosisYear(leftYear)).collect(Collectors.toList());
-            if (validRowsForLeftCode.isEmpty())
-                return MpResult.QUESTIONABLE;
-            for (HematoDbDTO dto : validRowsForLeftCode) {
-                if (dto.matches(rightCode, leftYear)) {
-                    rightCodeIsSamePrimaryForLeft = true;
-                    break;
-                }
-            }
-        }
-        if (rightCodeIsSamePrimaryForLeft && _samePrimaryDto.containsKey(rightCode)) {
-            List<HematoDbDTO> allRowsForRightCode = _samePrimaryDto.get(rightCode);
-            List<HematoDbDTO> validRowsForRightCode = allRowsForRightCode.stream().filter(r -> r.validCodeForDiagnosisYear(rightYear)).collect(Collectors.toList());
-            if (validRowsForRightCode.isEmpty())
-                return MpResult.QUESTIONABLE;
-            for (HematoDbDTO dto : validRowsForRightCode)
-                if (dto.matches(leftCode, rightYear))
-                    return MpResult.SINGLE_PRIMARY;
-        }
-        return MpResult.MULTIPLE_PRIMARIES;
+    public List<HematoDbDTO> getSamePrimary(String morphology) {
+        if (morphology == null || !_MORPHOLOGY.matcher(morphology).matches() || !_samePrimaryDto.containsKey(morphology))
```

Comment:
For this method and the 2 others, you should do two accesses to the map (one to check that the key is there and one to get the value). In most map implementation, checking the key is as expensive as getting the value (not that it's super expensive for a map, but still).

I think the most elegant way to do this (assuming that no key can be in the map and point to a null value) is to use the "Map.computeIfAbsent() method and pass the empty list as the second parameter. That's the most efficient way to not do two consecutive lookups in the map.